DEFEDIT-1163 Scene rendering performance improvements

ragnard-king opened the pull request:
### User facing changes

- Big scenes, especially with a large number of GUIs should now render
  much more quickly and be more pleasant to work with in editor.

### Technical changes

- scene rendering:
  - introduce a `:pass-overrides` renderable key that allows for
    overriding renderable attributes on a per-pass basis
  - change flatten-scene to not bash transients in place (unsafe)
  - better implementation of geom/aabb-transform which was a hot-spot
    in scene flattening
- gui rendering:
  - switch to vertex2 vbufs
  - improve render-lines and render-geom
  - use `:pass-overrides` to batch outline pass more effectively
- font rendering:
  - switch to vertex2 vbufs
  - remove unboxed math in hot code
  - cache and re-use vbufs if possible to reduce ByteBuffer allocation
    pressure
- label rendering:
  - adapt to font/gui changes

ragnard-king committed:
DEFEDIT-1163 Scene rendering performance improvements

* User facing changes

- Big scenes, especially with a large number of GUIs should now render
  much more quickly and be more pleasant to work with in editor.

* Technical changes

- scene rendering:
  - introduce a `:pass-overrides` renderable key that allows for
    overriding renderable attributes on a per-pass basis
  - change flatten-scene to not bash transients in place (unsafe)
  - better implementation of geom/aabb-transform which was a hot-spot
    in scene flattening
- gui rendering:
  - switch to vertex2 vbufs
  - improve render-lines and render-geom
  - use `:pass-overrides` to batch outline pass more effectively
- font rendering:
  - switch to vertex2 vbufs
  - remove unboxed math in hot code
  - cache and re-use vbufs if possible to reduce ByteBuffer allocation
    pressure
- label rendering:
  - adapt to font/gui changes

matgis-king reviewed (approved):
Found some minor things, but should be OK to merge as-is!

I also noticed we still get some boxed math warnings in these files if we put `(set! *unchecked-math* :warn-on-boxed)` at the top. But perhaps the bottleneck is elsewhere now?

ragnard-king committed:
DEFEDIT-1163 Review fixes

- if -> when for clarity
- minor improvement to gui/render-lines
- minor restructuring/renaming of scene and pass-renderables
  creation/updating/persiting

ragnard-king committed:
DEFEDIT-1163 Review fixes

- Use render-key specific to outline pass to ensure selected outlines
  are drawn on top of other outlines. We previously attempted to do
  this by sorting all renderables in outline pass when flattening the
  scene, but there's an additional sorting by `:render-key` that
  happens after that. Instead, use a pass specific render-key.
